cpp/src/quantiles/quantile.cu

```diff
@@ -13,102 +13,194 @@
 
-        auto selector = [&](size_type location) {
-            return get_array_value<T>(input_begin, column_order.ordering == order::ASCENDING ? location : -location);
-        };
+        thrust::transform(
+            q_device.begin(),
+            q_device.end(),
+            d_output->template begin<Result>(),
+            [sorted_data, interp=interp, size=size]
+            __device__ (double q){
+                return select_quantile_data<Result>(sorted_data, size, q, interp);
+            });
+
+        if (input.nullable())
+        {
+            auto sorted_validity = thrust::make_transform_iterator(
+                ordered_indices,
+                [input=*d_input] __device__ (size_type idx) {
+                    return input.is_valid_nocheck(idx);
+                });
+
+            rmm::device_buffer mask;
+            size_type null_count;
+
+            std::tie(mask, null_count) = valid_if(
+                q_device.begin(),
+                q_device.end(),
+                [sorted_validity, interp=interp, size=size]
+                __device__(double q) {
+                    return select_quantile_validity(sorted_validity,
+                                                    size,
+                                                    q,
+                                                    interp);
+                },
+                stream,
+                mr);
+
+            output->set_null_mask(std::move(mask), null_count);
+        }
 
-        auto result = select_quantile<ScalarResult>(selector, valid_count, q, interp);
-        return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
+        return output;
     }
 };
 
-} // anonymous namespace
+template<bool exact, typename SortMapIterator>
+std::unique_ptr<column>
+quantile(column_view const& input,
+         SortMapIterator ordered_indices,
+         size_type size,
+         std::vector<double> const& q,
+         interpolation interp,
+         bool retain_types,
+         rmm::mr::device_memory_resource* mr)
+{
+    auto functor = quantiles_functor<exact, SortMapIterator>{
+        ordered_indices,
+        size,
+        q,
+        interp,
+        retain_types,
+        mr,
+        0};
+
+    return type_dispatcher(input.type(), functor, input);
+}
+
 } // namespace detail
 
-std::unique_ptr<scalar>
+std::unique_ptr<column>
 quantile(column_view const& input,
-         double q,
+         std::vector<double> const& q,
          interpolation interp,
-         order_info column_order)
+         column_view const& ordered_indices,
+         bool exact,
+         rmm::mr::device_memory_resource* mr)
 {
-  CUDF_FUNC_RANGE();
-  return type_dispatcher(input.type(), detail::quantile_functor{}, input, q,
-                         interp, column_order);
+    CUDF_FUNC_RANGE();
+    CUDF_EXPECTS(input.size() > 0,
+                "quantiles requires at least one input row.");
+
+    if (ordered_indices.is_empty())
+    {
+        if (exact)
```

Comment:
This is nearly the same as `retain_types`, but it's only here for backwards-compatibility. I plan to move the discrete functionality to a new function, as discussed in #4292, in a followup PR.